### DISCUSSION TITLE
Custom snippet behaviour

### DISCUSSION THREAD

**Discussion by marbaa** [cobbler/cobbler#3005] [Q&A] [answered]
Hi,

was there some change in regards how custom snippet works in Cobbler 2.x and 3.x? I have some scripts (users, network configs, ssh configs, etc) which are loaded via ```$SNIPPET('custom/xxx')```. They works in 2.x. After installation, I can see all commands from custom snippet in ```anaconda-ks.cfg``` and ```original-ks.cfg```, but in fact, nothing happened.
I'm installing ```RHEL8.4``` from ```Cobbler 3.2.2``` installed on ```CentOS Stream release 8```.

**Comment by marbaa** [comment]
Difference between setups is, that with version 2.x I used manual editing of dhcp config, manual adding hostnames, ip, dns name. And one general system in cobbler. With 3.x, I'm trying to work with cobbler managed dhcp enabled, and multiple cobbler system with its own mac and ip and dns name to figure out which way is better. Next, the custom snippet scripts checks the $HOSTNAME returned from dhcp offer. So I think the problem is there and maybe, I'm using wrong parameter in cobbler system to set hostname.

**Reply by SchoolGuy** [reply]
So... I did try to fixup the logic of `$SNIPPET`, so there were bugs. However our current tests show that the inclusion like you do it should work. Do you see anything in `/var/log/cobbler/cobbler.log` which hints to the impossibility to resolve a variable?

**Comment by marbaa** [comment]
Nothing related to ```$SNIPPET```. Weird is this part of log where it says cannot find OS_disk:
```
2022-03-28T17:07:09 - INFO | REMOTE generate_autoinstall; user(?)
2022-03-28T17:07:09 - WARNING | errors were encountered rendering the template
2022-03-28T17:07:09 - WARNING |
[{'code': 'VFFSL(SL,"OS_disk",True)',
  'exc_val': NotFound("cannot find 'OS_disk'",),
  'lineCol': (47, 47),
  'rawCode': '${OS_disk}',
  'time': 'Mon Mar 28 17:07:09 2022'},
 {'code': 'VFFSL(SL,"OS_disk",True)',
  'exc_val': NotFound("cannot find 'OS_disk'",),
  'lineCol': (47, 47),
  'rawCode': '${OS_disk}',
  'time': 'Mon Mar 28 17:07:09 2022'},
 {'code': 'VFFSL(SL,"OS_disk",True)',
  'exc_val': NotFound("cannot find 'OS_disk'",),
  'lineCol': (47, 47),
  'rawCode': '${OS_disk}',
  'time': 'Mon Mar 28 17:07:09 2022'},
 {'code': 'VFFSL(SL,"OS_disk",True)',
  'exc_val': NotFound("cannot find 'OS_disk'",),
  'lineCol': (47, 47),
  'rawCode': '${OS_disk}',
  'time': 'Mon Mar 28 17:07:09 2022'},
 {'code': 'VFFSL(SL,"SNIPPET",False)(\'download_config_files\')',
  'exc_val': NotFound("cannot find 'replace' while searching for 'replace'",),
  'lineCol': (92, 1),
  'rawCode': "$SNIPPET('download_config_files')",
  'time': 'Mon Mar 28 17:07:09 2022'}]
  ```

I have this in rhel8.ks (part):
```
...
selinux --disabled
# Do not configure the X Window System
skipx
# System timezone
timezone  Europe/Berlin
# Clear the Master Boot Record
zerombr

%pre
if [[ `grep ": active" /proc/mdstat` != "" ]] ; then
        OS_disk=`cat /proc/mdstat |grep ": active" |awk '{ print $1 }'`
        echo "bootloader --location=mbr --driveorder=${OS_disk}" --append="ipv6.disable=1" > /tmp/bootloader.txt
        echo "clearpart --drives=${OS_disk} --all --initlabel" > /tmp/disk_selection.txt
else
        OS_disk=`lsblk -o NAME,SIZE | grep sd | grep G | head -n1 | awk '{ print $1 }'`
        echo "bootloader --location=mbr --driveorder=${OS_disk}" --append="ipv6.disable=1"> /tmp/bootloader.txt
        echo "clearpart --drives=${OS_disk} --all --initlabel" > /tmp/disk_selection.txt
fi

# exec < /dev/tty6 > /dev/tty6 2> /dev/tty6
# chvt 6
# read enter
# chvt 1
# exec < /dev/tty1> /dev/tty1 2> /dev/tty1

%end
$SNIPPET('custom/partitioning')

%pre
$SNIPPET('log_ks_pre')
$SNIPPET('autoinstall_start')
...
```

However when it boots into OS after installation, disks are partitionad according custom SNIPPET and bootloader is placed on proper disk. Custom partitioning snippet is outside any %pre/post %end section and ksvalidator gives this:
```
[root@cobblerVM templates]# ksvalidator rhel8.ks
The following problem occurred on line 61 of the kickstart file:

Unknown command: $SNIPPET(custom/partitioning)
```
However as I wrote, it installs correctly exactly as is written in partitioning. If I put that snippet into some %pre section, then ksvalidator runs fine, but installer hangs up, because of bad partitioning.

To put more confusion here 😂 Just now I found my other testing cobbler VM on second computer. This new cobbler installation is basically copy paste from second computer and that olders works without any problem. I don't get it, have to go through all config files line by line. DHCP, cobbler, snippets...

**Reply by SchoolGuy** [reply]
Oh you indeed have stuff related to `$SNIPPET`. Running the pure ksvalidator is not a good idea as `$SNIPPET` is not part of the kickstart syntax. For this we have the command [`cobbler validate-autoinstalls`](https://cobbler.readthedocs.io/en/latest/cobbler.html#cobbler-validate-autoinstalls). 

Long Story short: I will try to reproduce later with the version you gave above but I atm see no indication except a Cheetah Syntax error in the template why this shouldn't work.

**Comment by marbaa** [comment]
Still looking for mistakes, typos. I will try again copy kickstart and snippet files from older computer to newer.
Both cobbler's says it's ```3.2.2```, but with ```rpm -qa```, older computer says ```3.2.2-8```, and newer with problems ```3.2.2-10```.

**Reply by marbaa** [reply]
I've upgraded the working machine, and it keep working after it. So, I'm recreating whole virtual machine with cobbler. I was out of ideas 😞.

**Comment by marbaa** [comment, marked as answer]
Fresh new installation, same files copied as before and now everything works. I have no idea. 🤣 Thread can be closed, deleted.
